pydantic/generics.py

```diff
@@ -0,0 +1,87 @@
+from typing import Any, Dict, NoReturn, Tuple, Type, TypeVar, Union, get_type_hints
+
+from pydantic import BaseModel, create_model
+from pydantic.class_validators import gather_validators
+
+_generic_types_cache: Dict[Tuple[Type[Any], Union[Any, Tuple[Any, ...]]], Type[BaseModel]] = {}
+GenericModelT = TypeVar("GenericModelT", bound="GenericModel")
+
+
+class GenericModel(BaseModel):
+    __slots__ = ()
+    __parameters__ = ()
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> NoReturn:
+        if cls is GenericModel:
+            raise TypeError(f'Type {cls.__name__} cannot be instantiated; it can be used only as a base class')
+        else:
+            raise TypeError(f'Type {cls.__name__} cannot be instantiated without providing generic parameters')
+
+    def __class_getitem__(  # type: ignore
+        cls: Type[GenericModelT], params: Union[Any, Tuple[Any, ...]]
+    ) -> Type[BaseModel]:
+        if not isinstance(params, tuple):
+            params = (params,)
+        cached = _generic_types_cache.get((cls, params))
+        if cached is not None:
+            return cached
+
+        if cls is GenericModel:
+            raise TypeError(f'Type parameters should be placed on typing.Generic instead of GenericModel')
+
+        # Generate a parameterized BaseModel
+        check_parameters_count(cls, params)
+        typevars_map: Dict[Any, Any] = dict(zip(cls.__parameters__, params))
+        concrete_type_hints: Dict[str, Type[Any]] = {
+            k: resolve_type_hint(v, typevars_map) for k, v in get_type_hints(cls).items()
+        }
+
+        model_name = concrete_name(cls, params)
+        config = cls.Config if hasattr(cls, 'Config') else None
+        validators = gather_validators(cls)
+        fields: Dict[str, Tuple[Type[Any], Any]] = {
+            k: (v, getattr(cls, k, ...)) for k, v in concrete_type_hints.items()
+        }
+        created_model = create_model(
+            model_name=model_name,
+            __module__=cls.__module__,
+            __base__=None,
```

Comment:
I think I'm not enough of a pydantic pro to know whether it would ever make sense to specify `__base__` here. It seems much more likely that `__config__` would be useful. (I am explicitly specifying `__base__` here to get mypy to stop giving errors about the type of `fields`, which it strangely does if you skip an argument.)

Reply:
looks fine to me, as your test shows `Config` works which I think is sufficient.